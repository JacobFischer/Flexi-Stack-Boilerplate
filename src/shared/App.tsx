import React from "react";
import { Link, Route, Switch } from "react-router-dom";

export function App() {
    return (<>
        <ul>
            <li><Link to="/">Home</Link></li>
            <li><Link to="/about">About</Link></li>
            <li><Link to="/some-page">Some Page</Link></li>
            <li><Link to="/not-found">Dead link</Link></li>
        </ul>
        <Switch>
            <Route exact path="/" component={() => <p>Welcome to the index</p>} />
            <Route exact path="/about" component={() => <em>About the app</em>} />
            <Route exact path="/some-page" component={() => <strong>Some random other page</strong>} />
            <Route component={() => <pre>404 Not Found!</pre>} />
        </Switch>
    </>);
}
