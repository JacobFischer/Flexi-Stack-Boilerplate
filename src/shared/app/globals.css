/**
 * These are global styles that will be statically generated with a CSS reset,
 * and loaded with every page.
 * **Only** put styles in here that work on every rendered page, otherwise use
 * a styled-component to dynamically inject your styles when that component is
 * rendered
 */

body {
    background-color: lightgrey;
    font-size: 16px;
    padding: 1rem;
}
