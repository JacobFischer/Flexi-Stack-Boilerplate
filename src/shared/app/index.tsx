export { Body } from "./Body";
export { Head } from "./Head";
