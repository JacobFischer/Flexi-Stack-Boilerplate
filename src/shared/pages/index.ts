import * as allPages from './pages';
import * as pageNotFoundExports from './404';

export const pageNotFound = pageNotFoundExports;

export type PageExport = {
  title: string;
  route: string;
  Component: React.FunctionComponent;
};

export const pagesList: PageExport[] = Object.values(allPages).sort((a, b) =>
  a.route.localeCompare(b.route),
);
