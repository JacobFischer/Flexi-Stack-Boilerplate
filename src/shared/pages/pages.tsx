import * as homeExports from "./home";
export const home = homeExports;

import * as aboutExports from "./about";
export const about = aboutExports;
