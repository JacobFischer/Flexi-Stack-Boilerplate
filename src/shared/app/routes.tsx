import React from "react";
import { Route, StaticRouter, Switch } from "react-router-dom";
import { renderToStaticMarkup } from "react-dom/server";
import { pagesList, pageNotFound } from "../pages";

/**
 * Lol wut.
 *
 * @param props - Imagine dragons.
 * @param props.render - Js doc why.
 * @returns Stuff.
 */
export const Routes = <T extends React.ReactNode>(props: {
    render: (page: typeof pagesList[number], matchedRoute: boolean) => T;
}) => (
    <Switch>
        {pagesList.map((page) => (
            <Route
                key={page.route}
                exact
                path={page.route}
                render={() => props.render(page, true)}
            />
        ))}
        <Route render={() => props.render(pageNotFound, false)} />
    </Switch>
);

/**
 * Quickly checks if a location string matches any registered route.
 *
 * @param location - The location to check.
 * @returns True if a match was made, false otherwise.
 */
export function matchingRoute(location: string): boolean {
    const matchedRoute = renderToStaticMarkup(
        <StaticRouter location={location}>
            <Routes render={(_, matchedRoute) => String(matchedRoute)} />
        </StaticRouter>,
    );

    // boolean was transformed to a string via the quick react render above
    return matchedRoute == String(true);
}
