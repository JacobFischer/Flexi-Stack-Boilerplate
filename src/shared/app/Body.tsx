// This is the main entry point into the app
import React from "react";
import { Routes } from "./routes";

export const Body: React.FunctionComponent = () => (
    <Routes render={({ Component }) => <Component />} />
);
