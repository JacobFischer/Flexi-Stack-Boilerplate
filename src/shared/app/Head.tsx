// This is the main entry point into the app
import React from 'react';
import { Routes } from './routes';
import { useLocation, StaticRouter } from 'react-router-dom';
import { renderToStaticMarkup } from 'react-dom/server';
import globalsCss from './globals.css';

export const Head: React.FunctionComponent<{
  // TODO: this is kind of hack-y. react-helmet MUST have head children as
  // direct children,
  Wrapper?: React.ElementType;
}> = ({ Wrapper = React.Fragment }) => {
  // Ideally we could nest react-router route matches in the title
  // however, react-helmet **only** permits strings at the moment,
  // no other react nodes.
  // So, we'll just manually render out the title here to make it happy
  const title = renderToStaticMarkup(
    <StaticRouter location={useLocation()}>
      <Routes render={({ title }) => title}></Routes>
    </StaticRouter>,
  );

  return (
    <Wrapper>
      <link rel="stylesheet" type="text/css" href={globalsCss} media="all" />
      <title>{title}</title>
    </Wrapper>
  );
};
