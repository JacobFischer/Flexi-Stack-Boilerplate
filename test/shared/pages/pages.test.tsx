import React from 'react';
import { StaticRouter } from 'react-router';
import renderer from 'react-test-renderer';
import { pagesAnd404List } from '../../../src/shared/pages';
import { Body, Head } from '../../../src/shared/app';

describe('all pages', () => {
  pagesAnd404List.forEach((page) => {
    const { title, route, Component } = page;

    const render = (element: React.ReactElement) =>
      renderer.create(<StaticRouter location={route}>{element}</StaticRouter>);

    describe(`page '${title}'`, () => {
      it('should be the correct export shape', () => {
        expect(typeof page).toBe('object');
        expect(typeof Component).toBe('function');
        expect(typeof route).toBe('string');
        expect(typeof title).toBe('string');
      });

      it('renders', () => {
        expect(() => render(<Component />)).not.toThrow();
      });

      it('exists within the App', () => {
        const { root } = render(
          <html>
            <head>
              <Head />
            </head>
            <body>
              <Body />
            </body>
          </html>,
        );
        expect(root.findByType('title').children).toMatchObject([title]);
        expect(root.findAllByType(Component).length).toBe(1);
      });

      it('should match the last snapshot', () => {
        expect(render(<Component />).toJSON()).toMatchSnapshot();
      });
    });
  });
});
