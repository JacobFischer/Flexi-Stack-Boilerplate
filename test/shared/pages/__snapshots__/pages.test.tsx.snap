// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`all pages page '404 Not Found' renders 1`] = `
<div>
  404 Not Found
</div>
`;

exports[`all pages page '404 Not Found' renders as an App page 1`] = `
<html>
  <head>
    <title>
      404 Not Found
    </title>
  </head>
  <body>
    <div>
      404 Not Found
    </div>
  </body>
</html>
`;

exports[`all pages page 'About' renders 1`] = `
<div
  className="PageWrapper__Wrapper-sc-120esh3-0 dqljDC"
>
  <h1
    className="Header__DarkBlue-zt7fth-0 eCcsLu"
  >
    About
  </h1>
  <ul
    className="NavBar__List-pvivyq-0 JFDnC"
  >
    <li
      className="NavBar__Item-pvivyq-1 dvvGHr"
    >
      <a
        href="/"
        onClick={[Function]}
      >
        Home
      </a>
    </li>
    <li
      className="NavBar__Item-pvivyq-1 dvvGHr"
    >
      <a
        href="/about"
        onClick={[Function]}
      >
        About
      </a>
    </li>
    <li
      className="NavBar__Item-pvivyq-1 dvvGHr"
    >
      <a
        href="/nowhere"
        onClick={[Function]}
      >
        Nowhere
      </a>
    </li>
  </ul>
  <section>
    <div>
      About my boy
    </div>
  </section>
</div>
`;

exports[`all pages page 'About' renders as an App page 1`] = `
<html>
  <head>
    <title>
      About
    </title>
  </head>
  <body>
    <div
      className="PageWrapper__Wrapper-sc-120esh3-0 dqljDC"
    >
      <h1
        className="Header__DarkBlue-zt7fth-0 eCcsLu"
      >
        About
      </h1>
      <ul
        className="NavBar__List-pvivyq-0 JFDnC"
      >
        <li
          className="NavBar__Item-pvivyq-1 dvvGHr"
        >
          <a
            href="/"
            onClick={[Function]}
          >
            Home
          </a>
        </li>
        <li
          className="NavBar__Item-pvivyq-1 dvvGHr"
        >
          <a
            href="/about"
            onClick={[Function]}
          >
            About
          </a>
        </li>
        <li
          className="NavBar__Item-pvivyq-1 dvvGHr"
        >
          <a
            href="/nowhere"
            onClick={[Function]}
          >
            Nowhere
          </a>
        </li>
      </ul>
      <section>
        <div>
          About my boy
        </div>
      </section>
    </div>
  </body>
</html>
`;

exports[`all pages page 'Home Page' renders 1`] = `
<div
  className="PageWrapper__Wrapper-sc-120esh3-0 dqljDC"
>
  <h1
    className="Header__DarkBlue-zt7fth-0 eCcsLu"
  >
    Home Page
  </h1>
  <ul
    className="NavBar__List-pvivyq-0 JFDnC"
  >
    <li
      className="NavBar__Item-pvivyq-1 dvvGHr"
    >
      <a
        href="/"
        onClick={[Function]}
      >
        Home
      </a>
    </li>
    <li
      className="NavBar__Item-pvivyq-1 dvvGHr"
    >
      <a
        href="/about"
        onClick={[Function]}
      >
        About
      </a>
    </li>
    <li
      className="NavBar__Item-pvivyq-1 dvvGHr"
    >
      <a
        href="/nowhere"
        onClick={[Function]}
      >
        Nowhere
      </a>
    </li>
  </ul>
  <section>
    <div>
      Hello world from the app!
    </div>
  </section>
</div>
`;

exports[`all pages page 'Home Page' renders as an App page 1`] = `
<html>
  <head>
    <title>
      Home Page
    </title>
  </head>
  <body>
    <div
      className="PageWrapper__Wrapper-sc-120esh3-0 dqljDC"
    >
      <h1
        className="Header__DarkBlue-zt7fth-0 eCcsLu"
      >
        Home Page
      </h1>
      <ul
        className="NavBar__List-pvivyq-0 JFDnC"
      >
        <li
          className="NavBar__Item-pvivyq-1 dvvGHr"
        >
          <a
            href="/"
            onClick={[Function]}
          >
            Home
          </a>
        </li>
        <li
          className="NavBar__Item-pvivyq-1 dvvGHr"
        >
          <a
            href="/about"
            onClick={[Function]}
          >
            About
          </a>
        </li>
        <li
          className="NavBar__Item-pvivyq-1 dvvGHr"
        >
          <a
            href="/nowhere"
            onClick={[Function]}
          >
            Nowhere
          </a>
        </li>
      </ul>
      <section>
        <div>
          Hello world from the app!
        </div>
      </section>
    </div>
  </body>
</html>
`;
